File: flutter/platform/lib/src/modules/sync.dart
import '../core/http_client.dart';
import '../core/api_endpoints.dart';
import '../core/result.dart';

/// Sync module — user data push/pull and key-level CRUD.
class SyncModule {
  const SyncModule();

  /// Get sync status.
  Future<ApiResult<Map<String, dynamic>>> getStatus() =>
      HttpClient.instance.getMap(ApiEndpoints.syncStatus);

  /// Push data (full replace).
  Future<ApiResult<Map<String, dynamic>>> push(Map<String, dynamic> data) =>
      HttpClient.instance.postMap(ApiEndpoints.syncPush, body: data);

  /// Pull data.
  Future<ApiResult<Map<String, dynamic>>> pull() =>
      HttpClient.instance.getMap(ApiEndpoints.syncPull);

  /// Get all stored data keys.
  Future<ApiResult<Map<String, dynamic>>> getAllData() =>
      HttpClient.instance.getMap(ApiEndpoints.syncData);

  /// Get a single data key.
  Future<ApiResult<Map<String, dynamic>>> getDataKey(String key) =>
      HttpClient.instance.getMap(ApiEndpoints.syncDataKey(key));

  /// Update a single data key (full replace).
  Future<ApiResult<Map<String, dynamic>>> putDataKey(String key, Map<String, dynamic> data) =>
      HttpClient.instance.putMap(ApiEndpoints.syncDataKey(key), body: data);

  /// Patch a single data key (delta merge/remove).
  Future<ApiResult<Map<String, dynamic>>> patchDataKey(
    String key, {
    Map<String, dynamic>? merge,
    List<String>? remove,
  }) => HttpClient.instance.patchMap(ApiEndpoints.syncDataKey(key), body: {
    if (merge != null && merge.isNotEmpty) 'merge': merge,
    if (remove != null && remove.isNotEmpty) 'remove': remove,
  });

  /// Delete a single data key.
  Future<ApiResult<void>> deleteDataKey(String key) =>
      HttpClient.instance.deleteVoid(ApiEndpoints.syncDataKey(key));
}

File: web/platform/src/modules/sync.ts
import { http } from "../core/http";
import { Endpoints } from "../core/endpoints";
import type { ApiResult } from "../core/result";

export const sync = {
  getStatus(): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.syncStatus);
  },
  push(
    data: Record<string, unknown>
  ): Promise<ApiResult<Record<string, unknown>>> {
    return http.postMap(Endpoints.syncPush, data);
  },
  pull(): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.syncPull);
  },
  getAllData(): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.syncData);
  },
  getDataKey(key: string): Promise<ApiResult<Record<string, unknown>>> {
    return http.getMap(Endpoints.syncDataKey(key));
  },
  putDataKey(
    key: string,
    data: Record<string, unknown>
  ): Promise<ApiResult<Record<string, unknown>>> {
    return http.putMap(Endpoints.syncDataKey(key), data);
  },
  patchDataKey(
    key: string,
    body: { merge?: Record<string, unknown>; remove?: string[] }
  ): Promise<ApiResult<Record<string, unknown>>> {
    return http.patchMap(Endpoints.syncDataKey(key), body);
  },
  deleteDataKey(key: string): Promise<ApiResult<null>> {
    return http.deleteVoid(Endpoints.syncDataKey(key));
  },
};
